debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/AbstractLogMinerEventProcessor.java

```diff
@@ -221,36 +221,74 @@ protected Instant getLastProcessedScnChangeTime() {
      * @param row the event row that contains the row identifier, must not be {@code null}
      */
     protected void removeEventWithRowId(LogMinerEventRow row) {
-        // locate the events based solely on XIDUSN and XIDSLT.
-        String basePrefix = getTransactionIdPrefix(row.getTransactionId());
-        List<String> eventKeysForBasePrefix = getTransactionKeysWithPrefix(basePrefix);
-
-        String transactionIdPrefix = row.getTransactionId() + "-";
-
-        // filter the existing list down to the events for the transaction
-        List<String> eventKeys = eventKeysForBasePrefix.stream()
-                .filter(k -> k.startsWith(transactionIdPrefix))
-                .toList();
-
-        if (eventKeys.isEmpty() && isTransactionIdWithNoSequence(row.getTransactionId())) {
+        final T transaction = getTransactionCache().get(row.getTransactionId());
+        if (transaction != null) {
+            if (removeTransactionEventWithRowId(transaction, row)) {
+                return;
+            }
+            Loggings.logWarningAndTraceRecord(LOGGER, row,
+                    "Cannot undo change on table '{}' since event with row-id {} was not found.",
+                    row.getTableId(), row.getRowId());
+        }
+        else if (isTransactionIdWithNoSequence(row.getTransactionId())) {
             // This means that Oracle LogMiner found an event that should be undone but its corresponding
             // undo entry was read in a prior mining session and the transaction's sequence could not be
             // resolved.
-
+            final String prefix = getTransactionIdPrefix(row.getTransactionId());
             LOGGER.debug("Undo change refers to a transaction that has no explicit sequence, '{}'", row.getTransactionId());
-            LOGGER.debug("Checking all transactions with prefix '{}'", basePrefix);
-            eventKeys = eventKeysForBasePrefix;
-        }
+            LOGGER.debug("Checking all transactions with prefix '{}'", prefix);
+
+            final List<T> transactions = getTransactionCache().streamAndReturn(
+                    stream -> stream.filter(entry -> entry.getKey().startsWith(prefix))
+                            .map(LogMinerCache.Entry::getValue)
+                            .toList());
```

Comment:
If you wanted to avoid streaming across the entire transaction cache and creating the list, you could do the following:

`
stream.filter(entry -> entry.getKey().startsWith(prefix))
                            .map(LogMinerCache.Entry::getValue)
                            .anyMatch(t -> removeTransactionEventWithRowId(t, row))

`

if false is reported, you could then log the cannot undo message.
Curious if this would gain any performance in your test.

Reply:
Thanks, that certainly looks cleaner to me.  I will take a peek and see if it changes the performance.

Reply:
I don't see a significant change in the performance, but it may have better performance with a cache that contains more than one active transaction like my tests.